File: src/bindings-python/tests/security_test.py
"""Detailed validation smoke tests for every security fixture."""

import multiprocessing
import os
import queue
import unittest

from cloudformation_validate import CelEngine, RegoEngine, Severity, ValidateConfig, ValidationError

TESTS_DIR = os.path.dirname(os.path.abspath(__file__))
WORKSPACE = os.path.dirname(os.path.dirname(TESTS_DIR))
SECURITY_ROOT = os.path.join(WORKSPACE, "resources", "security")
SECURITY_TIMEOUT_SECONDS = 120


def discover_security_templates():
    templates = []
    for directory, _, filenames in os.walk(SECURITY_ROOT):
        for filename in filenames:
            if filename.endswith((".json", ".yaml", ".yml")):
                templates.append(os.path.join(directory, filename))
    return sorted(templates)


def validate_security_template(engine_name, template_path, outcome_queue):
    try:
        engine = RegoEngine() if engine_name == "rego" else CelEngine()
        config = ValidateConfig(severity_level=Severity.DEBUG)
        report = engine.validate_detailed(template_path, config)
        if report.status is None or not isinstance(report.diagnostics, list):
            outcome_queue.put(("error", "detailed validation returned an incomplete report"))
            return
        outcome_queue.put(("ok", ""))
    except ValidationError as error:
        outcome_queue.put(("structured_error", str(error)))
    except BaseException as error:  # The child must communicate unexpected binding failures.
        outcome_queue.put(("error", repr(error)))


class SecurityTemplateTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.templates = discover_security_templates()
        if not cls.templates:
            raise AssertionError(f"no security templates found under {SECURITY_ROOT}")
        cls.process_context = multiprocessing.get_context("spawn")

    def test_every_security_template_with_both_engines(self):
        for engine_name in ("rego", "cel"):
            for template_path in self.templates:
                relative_path = os.path.relpath(template_path, SECURITY_ROOT).replace(os.sep, "/")
                with self.subTest(engine=engine_name, template=relative_path):
                    outcome_queue = self.process_context.Queue()
                    process = self.process_context.Process(
                        target=validate_security_template,
                        args=(engine_name, template_path, outcome_queue),
                    )
                    process.start()
                    process.join(SECURITY_TIMEOUT_SECONDS)
                    if process.is_alive():
                        process.terminate()
                        process.join(5)
                        if process.is_alive():
                            process.kill()
                            process.join()
                        self.fail(
                            f"{engine_name}/{relative_path} exceeded the hard "
                            f"{SECURITY_TIMEOUT_SECONDS}-second limit"
                        )
                    try:
                        status, message = outcome_queue.get(timeout=5)
                    except queue.Empty:
                        self.fail(
                            f"{engine_name}/{relative_path} exited with status {process.exitcode} "
                            "without returning an outcome"
                        )
                    if relative_path == "deep_nesting.json" and status == "structured_error":
                        self.assertTrue(message, "deep nesting must return a structured error")
                        continue
                    self.assertEqual("ok", status, f"{engine_name}/{relative_path}: {message}")


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: src/bindings-jvm/tests/kotlin/src/test/kotlin/SecurityTest.kt
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertTimeoutPreemptively
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.TestFactory
import software.amazon.cloudformation.validate.CelEngine
import software.amazon.cloudformation.validate.RegoEngine
import software.amazon.cloudformation.validate.ValidateConfig
import software.amazon.cloudformation.validate.ValidationException
import software.amazon.cloudformation.validate.engine.EngineConfig
import software.amazon.cloudformation.validate.rules.Severity
import java.io.File
import java.time.Duration

class SecurityTest {
    @TestFactory
    fun everySecurityTemplateWithBothEngines(): List<DynamicTest> {
        val templates =
            securityRoot
                .walkTopDown()
                .filter { it.isFile && it.extension.lowercase() in setOf("json", "yaml", "yml") }
                .sortedBy { it.path }
                .toList()
        check(templates.isNotEmpty()) { "no security templates found under ${securityRoot.path}" }

        val config = ValidateConfig(severityLevel = Severity.DEBUG)
        return listOf("rego", "cel").flatMap { engineName ->
            templates.map { template ->
                val relativePath = template.relativeTo(securityRoot).path.replace('\\', '/')
                DynamicTest.dynamicTest("$engineName/$relativePath") {
                    assertTimeoutPreemptively(Duration.ofSeconds(60)) {
                        val engine: Any =
                            if (engineName == "rego") {
                                RegoEngine(EngineConfig())
                            } else {
                                CelEngine(EngineConfig())
                            }
                        try {
                            val report =
                                when (engine) {
                                    is RegoEngine -> engine.validateDetailed(template, config)
                                    is CelEngine -> engine.validateDetailed(template, config)
                                    else -> error("unsupported engine type")
                                }
                            assertNotNull(report.status)
                            assertNotNull(report.diagnostics)
                        } catch (error: ValidationException) {
                            assertTrue(
                                relativePath == "deep_nesting.json" && !error.message.isNullOrEmpty(),
                                "$engineName/$relativePath returned an unexpected error: ${error.message}",
                            )
                        }
                    }
                }
            }
        }
    }

    companion object {
        private val resourcesRoot =
            listOf(
                File("${System.getProperty("user.dir")}/../../resources"),
                File("${System.getProperty("user.dir")}/../../../resources"),
            ).first { it.exists() }
        private val securityRoot = File(resourcesRoot, "security")
    }
}
